src/test/groovy/graphql/execution/AsyncFetchThenCompleteExecutionStrategy.java

```diff
@@ -0,0 +1,105 @@
+
+        Map<String, CompletableFuture<Object>> fetchFutures = new LinkedHashMap<>();
+
+        // first fetch every value
+        for (String fieldName : fields.keySet()) {
+            List<Field> fieldList = fields.get(fieldName);
+
+            ExecutionStrategyParameters newParameters = newParameters(parameters, fieldName);
+
+            CompletableFuture<Object> fetchFuture = supplyAsync(() ->
+                    fetchField(executionContext, newParameters, fieldList), executor);
+            fetchFutures.put(fieldName, fetchFuture);
+        }
+
+        // right here would be a good point to run a dataloader.dispatch() say to ensure batching of field fetches
+        // we would need a call out mechanism, either an overloadable method or passed in call back
+
+        // now wait for all fetches to finish together via this join
+        allOf(fetchFutures.values()).join();
+
+        Map<String, Object> fetchedValues = new LinkedHashMap<>();
+        fetchFutures.forEach((k, v) -> fetchedValues.put(k, v.join()));
+        return fetchedValues;
+    }
+
+    private ExecutionResult completeFields(ExecutionContext executionContext, ExecutionStrategyParameters parameters, Map<String, Object> fetchedValues) {
+        Map<String, List<Field>> fields = parameters.fields();
+
+        // then for every fetched value, complete it, breath first
+        Map<String, Object> results = new LinkedHashMap<>();
+        for (String fieldName : fetchedValues.keySet()) {
+            List<Field> fieldList = fields.get(fieldName);
+
+            ExecutionStrategyParameters newParameters = newParameters(parameters, fieldName);
+
+            Object fetchedValue = fetchedValues.get(fieldName);
+            try {
+                ExecutionResult resolvedResult = completeField(executionContext, newParameters, fieldList, fetchedValue);
+                results.put(fieldName, resolvedResult != null ? resolvedResult.getData() : null);
+            } catch (NonNullableFieldWasNullException e) {
+                assertNonNullFieldPrecondition(e);
+                results = null;
+                break;
+            }
+        }
+        return new ExecutionResultImpl(results, executionContext.getErrors());
+    }
+
+    private ExecutionStrategyParameters newParameters(ExecutionStrategyParameters parameters, String fieldName) {
+        ExecutionPath fieldPath = parameters.path().segment(fieldName);
+        return parameters.transform(builder -> builder.path(fieldPath));
+    }
+
+
+    public static <T> CompletableFuture<List<T>> allOf(final Collection<CompletableFuture<T>> futures) {
+        CompletableFuture[] cfs = futures.toArray(new CompletableFuture[futures.size()]);
+
+        return CompletableFuture.allOf(cfs)
+                .thenApply(vd -> futures.stream()
+                        .map(CompletableFuture::join)
+                        .collect(Collectors.toList())
+                );
+    }
+}
```

Comment:
This could be the general basis for a future "async" version that is breath first like the reference implementation